Pod/Classes/LSTracer.m

```diff
@@ -348,42 +348,55 @@ - (void)flush:(void (^)(NSError *_Nullable error))doneCallback {
-                   completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
-                       @try {
-                           __typeof(self) strongSelf = weakSelf;
-                           SInt64 destinationMicros = [LSClockState nowMicros];
-                           NSError *jsonError;
-                           NSDictionary *responseJSON =
-                               [NSJSONSerialization JSONObjectWithData:data options:kNilOptions error:&jsonError];
-                           if (jsonError == nil) {
-                               if ([responseJSON objectForKey:@"timing"] != nil) {
-                                   NSDictionary *timingJSON = [responseJSON objectForKey:@"timing"];
-                                   NSNumber *receiveMicros = [timingJSON objectForKey:@"receive_micros"];
-                                   NSNumber *transmitMicros = [timingJSON objectForKey:@"transmit_micros"];
-
-                                   if (receiveMicros != nil && transmitMicros != nil) {
-                                       // Update our local NTP-lite clock state with the latest
-                                       // measurements.
-                                       [strongSelf.clockState addSampleWithOriginMicros:originMicros
-                                                                          receiveMicros:receiveMicros.longLongValue
-                                                                         transmitMicros:transmitMicros.longLongValue
-                                                                      destinationMicros:destinationMicros];
-                                   }
-                               }
-                           }
-                       } @catch (NSException *e) {
-                           NSLog(@"Caught exception in LightStep reporting response; dropping "
-                                 @"data. Exception: %@",
-                                 e);
-                       } @finally {
-                           cleanupBlock(true, error);
-                       }
-                   }];
-    // "Start" (resume) the HTTP activity.
-    [postDataTask resume];
+    SInt64 originMicroseconds = [[self.timeProvider currentTime] toMicros];
+    [[session dataTaskWithRequest:request
+                completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+                    // Hack, but better than before
+                    if (![self.timeProvider isKindOfClass:[LSClockState class]) {
+                        cleanupBlock(true, error);
+                        return
+                    }
+
+                    [self handleAPIResponse:data
+                     withOriginMicroseconds:originMicroseconds
+                                 completion:cleanupBlock
+                                      error:error];
+                }]
+     resume];
+}
+
+- (void)handleAPIResponse:(NSData *)data
```

Comment:
right, just moved since the upper method was getting hard to follow with added branching